src/main/java/io/vertx/core/http/impl/Http2HeadersAdaptor.java

```diff
@@ -95,10 +96,14 @@ public int size() {
 
   @Override
   public List<Map.Entry<String, String>> entries() {
-    return headers.names()
-        .stream()
-        .map(name -> new AbstractMap.SimpleEntry<>(name.toString(), headers.get(name).toString()))
-        .collect(Collectors.toList());
+    if (isEmpty()) {
+      return Collections.emptyList();
+    }
+    List<Map.Entry<String, String>> list = new ArrayList<Map.Entry<String, String>>(headers.size());
+    for (Map.Entry<String, String> entry : this) {
```

Comment:
I think dup entries are required. Otherwise, ClientConnection.createRequest will not work correctly.